zilencer/management/commands/calculate_first_visible_message_id.py

```diff
@@ -0,0 +1,35 @@
+from typing import Any
+
+from django.core.management.base import CommandParser
+
+from zerver.lib.management import ZulipBaseCommand
+from zerver.lib.cache import realm_first_visible_message_id_cache_key, \
+    cache_get, cache_set
+from zerver.lib.message import estimiate_recent_messages
+from zerver.models import Realm, Message
+
+
+class Command(ZulipBaseCommand):
+    help = """Calculate the value of first visible message ID and store it in cache"""
+
+    def add_arguments(self, parser: CommandParser) -> None:
+        self.add_realm_args(parser)
+
+    def handle(self, *args: Any, **options: str) -> None:
+        realm = self.get_realm(options)
+
+        if realm is None:
+            realms = Realm.objects.all()
+        else:
+            realms = [realm]
+        for realm in realms:
+            cache_expired = cache_get(realm_first_visible_message_id_cache_key(realm)) is None
+            if realm.message_visibility_limit is not None and \
+                    (estimiate_recent_messages(realm, 1) > 0 or cache_expired):
+                try:
+                    first_visible_message_id = Message.objects.filter(sender__realm=realm).values('id').\
+                        order_by('-id')[realm.message_visibility_limit - 1]["id"]
+                except IndexError:
+                    first_visible_message_id = 0
+                cache_set(realm_first_visible_message_id_cache_key(realm),
+                          first_visible_message_id, timeout=3600*24*7)
```

Comment:
In general I think it's best to keep as much of the interesting logic as possible in library functions, squeezing it out of management commands so that they're left as just thin shells of command-line parsing. That's good for at least two things:
* testing;
* reusing the logic, either in writing more code later or from `manage.py shell`.

In this case, I think that looks like
* The body of this for-loop reduces to just a function call, perhaps `maybe_update_first_visible_message_id(realm, lookback_hours=30)`. (The lookback window corresponds to the period of the cron job, so it makes sense to pass it here. Though actually, even better would be to pass it down from the cron job as an argument to this command.)
* That function with "maybe" at the start of its name contains the logic up through this "if"-condition to decide whether to do the expensive thing. Then the body of the "if" is another function call that unconditionally does the expensive thing, like `update_first_visible_message_id(realm)`.
* That latter function does the `Message` query and stores the result.

With that structure, you can write nice unit tests for the library functions, including tests that look a lot like the tests from the previous version. And I think the command tests become a lot simpler -- just testing the command-line parsing.